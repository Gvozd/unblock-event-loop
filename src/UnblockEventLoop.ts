import { EventEmitter } from 'node:events';
import { AsyncQueue } from './AsyncQueue';

export class UnblockEventLoop extends EventEmitter{
    constructor(private threshold: number = 1) {
        super();
        this.queue = new AsyncQueue(this.queueExecutor);
    }

    unblock = () => {
        return new Promise((resolve) => {
            this.queue.push(resolve);
        });
    }

    private queue: AsyncQueue<(data: void) => void, void>;
    private start: number | null = null;
    private timer: Promise<void> | null = null;

    private queueExecutor = async (next: (data: void) => void) => {
        if (!this.timer) {
            this.timer = new Promise((resolve) => {
                setImmediate(() => {
                    this.emit('loop');
                    this.start = null;
                    this.timer = null;
                    resolve();
                });
            });
        }
        if (!this.start) {
            this.start = performance.now();
        }

        next(void 0);
        await Promise.resolve();

        const end = performance.now();
        if (end - this.start > this.threshold) {
            await this.timer;
        }
    }
}
