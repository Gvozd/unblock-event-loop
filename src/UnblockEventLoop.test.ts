import {UnblockEventLoop} from './UnblockEventLoop';
import {setTimeout as delay} from 'timers/promises';

describe('EventLoopMonitor', () => {
    const jestConsole = console;
    beforeAll(() => {
        global.console = require('console');
    });
    afterAll(() => {
        global.console = jestConsole;
    });
    it('base', async () => {
        const steps = [];
        steps.push('before start');
        const unblockEventLoop = new UnblockEventLoop();
        unblockEventLoop.addListener('loop', () => steps.push('loop'));
        const promises = Array(10).fill(void 0).map(async(el, idx) => {
            steps.push(`beforeLock: ${idx}`);
            await unblockEventLoop.unblock();
            steps.push(`afterLock: ${idx}`);
            var time = performance.now();
            while(performance.now() - time < 1);
            steps.push(`afterExecution: ${idx}`);
            return idx;
        });
        const results = await Promise.all(promises);
        expect(results, 'correct result').toEqual([0, 1, 2, 3, 4, 5, 6, 7, 8, 9]);
        expect(steps, 'execution order').toMatchSnapshot();
        await delay(0);
    });
    it('task asynchronously blocking loop', async () => {
        const steps = [];
        steps.push('before start');
        const unblockEventLoop = new UnblockEventLoop();
        unblockEventLoop.addListener('loop', () => steps.push('loop'));
        const promises = Array(10).fill(void 0).map(async(el, idx) => {
            steps.push(`beforeLock: ${idx}`);
            await unblockEventLoop.unblock();
            steps.push(`afterLock: ${idx}`);
            var time = performance.now()
            while(performance.now() - time < 1) await void 0;
            steps.push(`afterExecution: ${idx}`);
            return idx;
        });
        const results = await Promise.all(promises);
        expect(results, 'correct result').toEqual([0, 1, 2, 3, 4, 5, 6, 7, 8, 9]);
        expect(steps, 'execution order').toMatchSnapshot();
        await delay(0);
    });
});
