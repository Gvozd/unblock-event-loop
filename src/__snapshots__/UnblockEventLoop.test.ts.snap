// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EventLoopMonitor base 1`] = `
[
  "before start",
  "beforeLock: 0",
  "beforeLock: 1",
  "beforeLock: 2",
  "beforeLock: 3",
  "beforeLock: 4",
  "beforeLock: 5",
  "beforeLock: 6",
  "beforeLock: 7",
  "beforeLock: 8",
  "beforeLock: 9",
  "afterLock: 0",
  "afterExecution: 0",
  "loop",
  "afterLock: 1",
  "afterExecution: 1",
  "loop",
  "afterLock: 2",
  "afterExecution: 2",
  "loop",
  "afterLock: 3",
  "afterExecution: 3",
  "loop",
  "afterLock: 4",
  "afterExecution: 4",
  "loop",
  "afterLock: 5",
  "afterExecution: 5",
  "loop",
  "afterLock: 6",
  "afterExecution: 6",
  "loop",
  "afterLock: 7",
  "afterExecution: 7",
  "loop",
  "afterLock: 8",
  "afterExecution: 8",
  "loop",
  "afterLock: 9",
  "afterExecution: 9",
]
`;
